lightstep/sdk/metric/internal/syncstate/sync.go

```diff
@@ -0,0 +1,212 @@
+// updates, and the number of collected updates.
+type record struct {
+	refMapped refcountMapped
+
+	// updateCount is incremented on every Update.
+	updateCount int64
+
+	// collectedCount is set to updateCount on collection,
+	// supports checking for no updates during a round.
+	collectedCount int64
+
+	// accumulator can be a multi-accumulator if there
+	// are multiple behaviors or multiple readers, but
+	// these distinctions are not relevant for synchronous
+	// instruments.
+	accumulator viewstate.Accumulator
+}
+
+// snapshotAndProcess checks whether the accumulator has been
+// modified since the last collection (by any reader), returns a
+// boolean indicating whether the record is active.  If active, calls
+// SnapshotAndProcess on the associated accumulator and returns true.
+// If updates happened since the last collection (by any reader),
+// returns false.
+func (rec *record) snapshotAndProcess(final bool) bool {
+	mods := atomic.LoadInt64(&rec.updateCount)
+	coll := atomic.LoadInt64(&rec.collectedCount)
+
+	if mods == coll {
+		return false
+	}
+	rec.accumulator.SnapshotAndProcess(final)
+
+	// Updates happened in this interval, collect and continue.
+	atomic.StoreInt64(&rec.collectedCount, mods)
+	return true
+}
+
+// capture performs a single update for any synchronous instrument.
+func capture[N number.Any, Traits number.Traits[N]](_ context.Context, inst *Instrument, num N, attrs []attribute.KeyValue) {
+	if inst == nil {
+		// Instrument was completely disabled by the view.
+		return
+	}
+
+	// Note: Here, this is the place to use context, e.g., extract baggage.
+
+	if !aggregator.RangeTest[N, Traits](num, inst.descriptor.Kind) {
+		return
+	}
+
+	rec, updater := acquireRecord[N](inst, attrs)
+	defer rec.refMapped.unref()
+
+	updater.Update(num)
+
+	// Record was modified.
+	atomic.AddInt64(&rec.updateCount, 1)
+}
+
+// acquireRecord gets or creates a `*record` corresponding to `attrs`,
+// the input attributes.
+func acquireRecord[N number.Any](inst *Instrument, attrs []attribute.KeyValue) (*record, viewstate.Updater[N]) {
+	aset := attribute.NewSet(attrs...)
+
+	if lookup, ok := inst.current.Load(aset); ok {
+		// Existing record case.
+		rec := lookup.(*record)
+
+		if rec.refMapped.ref() {
+			// At this moment it is guaranteed that the
+			// record is in the map and will not be removed.
+			return rec, rec.accumulator.(viewstate.Updater[N])
+		}
+		// This record is no longer mapped, try to add a new
+		// record below.
+	}
+
+	// Note: the accumulator set below is created speculatively;
+	// if it is never returned, it will not be updated and can be
+	// safely discarded.
+	newRec := &record{
+		refMapped:   newRefcountMapped(),
+		accumulator: inst.compiled.NewAccumulator(aset),
+	}
+
+	for {
+		if found, loaded := inst.current.LoadOrStore(aset, newRec); loaded {
+			oldRec := found.(*record)
+			if oldRec.refMapped.ref() {
+				return oldRec, oldRec.accumulator.(viewstate.Updater[N])
+			}
+			runtime.Gosched()
```

Comment:
is this essentially telling the compiler to wait until the old record is removed before a new one can be added?

Reply:
Yes. There was a comment around `Delete()` on this topic, and I've added more here:
			// When this happens, we are waiting for the call to Delete()
			// inside SnapshotAndProcess() to complete before inserting
			// a new record.  This avoids busy-waiting.

(This design is lifted from the old SDK.)